### PR TITLE
feature/persistance-giftCertificate

### PR DESCRIPTION
**Pull request by AlekseySayarkin**

### CODE REVIEW

File: persistence/build.gradle [hunk 1]
```diff
@@ -10,5 +10,11 @@ repositories {
 }
 
 dependencies {
-    testCompile group: 'junit', name: 'junit', version: '4.12'
+    compile project(":model")
+    compile group: 'org.springframework', name: 'spring-context', version: '5.2.9.RELEASE'
```
Comment (churaalena) [inline persistence/build.gradle:14]: Please remove this dependencies

File: persistence/src/main/java/com/epam/esm/dao/GiftCertificateDAO.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package com.epam.esm.dao;
+
+import com.epam.esm.model.GiftCertificate;
+
+import java.util.List;
+
+public interface GiftCertificateDAO {
+
+    GiftCertificate getGiftCertificate(String name);
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/GiftCertificateDAO.java:9]: Is name unique per schema?

File: persistence/src/main/java/com/epam/esm/dao/GiftCertificateDAO.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    GiftCertificate getGiftCertificate(int id);
+    List<GiftCertificate> getAllGiftCertificates();
+    List<GiftCertificate> getGiftCertificateByTAgName(String tagName);
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/GiftCertificateDAO.java:12]: Please check spelling

File: persistence/src/main/java/com/epam/esm/dao/GiftCertificateDAO.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+    int addGiftCertificate(GiftCertificate giftCertificate);
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/GiftCertificateDAO.java:14]: Please check file formating

File: persistence/src/main/java/com/epam/esm/dao/extractor/GiftCertificateExtractor.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                map.put(giftCertificate.getId(), giftCertificate);
+            }
+
+            initTags(rs, giftCertificate);
+        }
+
+        return map;
+    }
+
+    private GiftCertificate initGiftCertificate(ResultSet rs) throws SQLException {
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/extractor/GiftCertificateExtractor.java:42]: Its not a init method. Please check method names

File: persistence/src/main/java/com/epam/esm/dao/extractor/GiftCertificateExtractor.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+@Component
+public class GiftCertificateExtractor implements ResultSetExtractor<List<GiftCertificate>> {
+
+    @Override
+    public List<GiftCertificate> extractData(ResultSet rs) throws SQLException {
+        return new ArrayList<>(extractDataFromResultSet(rs).values());
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/extractor/GiftCertificateExtractor.java:23]: Why we can`t extract List not a map?

File: persistence/src/main/java/com/epam/esm/dao/extractor/GiftCertificateExtractor.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    private Map<Integer, GiftCertificate> extractDataFromResultSet(ResultSet rs) throws SQLException {
+        Map<Integer, GiftCertificate> map = new LinkedHashMap<>();
+
+        while (rs.next()) {
+            GiftCertificate giftCertificate = map.get(rs.getInt(1));
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/extractor/GiftCertificateExtractor.java:30]: Please avoid magic numbers

File: persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    private static final String SQL_DELETE_CERTIFICATE =
+            "delete from GiftCertificates where (ID = ?)";
+
+    private static final String SQL_DELETE_JOIN =
+            "delete from CertificateDetails where (CertificateID = ?)";
+
+    private static final String SQL_UPDATE_CERTIFICATE =
+            "update GiftCertificates " +
+            "set Name = ?, Description = ?," +
+            "Price = ?, CreateDate = ?, LastUpdateDate = ?, Duration = ? " +
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:64]: Why do we need to update CreateDate here?

File: persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+            "where (ID = ?)";
+
+    private static final String SQL_GET_CERTIFICATE_BY_TAG_NAME =
+            "select * from GiftCertificates " +
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:68]: Please simplify this query without exist

File: persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+    @Override
+    public List<GiftCertificate> getGiftCertificateByTAgName(String tagName) {
+        return jdbcTemplate.query(SQL_GET_CERTIFICATE_BY_TAG_NAME, extractor, tagName);
+    }
+
+    @Override
+    public int addGiftCertificate(GiftCertificate giftCertificate) {
+        int id = addGiftCertificateToDB(giftCertificate);
+        if (id == 0) {
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:125]: Why id can be 0?

File: persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+            return 0;
+        }
+
+        addNotExistingTagsToDB(giftCertificate.getTags());
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:129]: Why not to combine this two loops?

File: persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+                Timestamp.from(ZonedDateTime.ofInstant(Instant.now(), ZoneOffset.UTC).toInstant()),
+                giftCertificate.getDuration(), giftCertificate.getId()
+        );
+
+        updateTags(giftCertificate, previous);
+
+        return true;
+    }
+
+    private void setNullFields(GiftCertificate giftCertificate, GiftCertificate previous) {
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:189]: Please rename method

File: persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    @Override
+    public boolean updateGiftCertificate(GiftCertificate giftCertificate) {
+        GiftCertificate previous = getGiftCertificate(giftCertificate.getId());
+
+        setNullFields(giftCertificate, previous);
+
+        jdbcTemplate.update(
+                SQL_UPDATE_CERTIFICATE, giftCertificate.getName(), giftCertificate.getDescription(),
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:178]: Why not to create some param map?

File: web/src/main/java/com.epam.esm/config/AppConfig.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.PropertySource;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.ResultSetExtractor;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.jdbc.datasource.DriverManagerDataSource;
+
+import javax.sql.DataSource;
+import java.util.List;
+
```
Comment (churaalena) [inline web/src/main/java/com.epam.esm/config/AppConfig.java:21]: Please, add comments to all classes

File: persistence/src/main/java/com/epam/esm/dao/GiftCertificateDAO.java (hunk 2 above)
Comment (AlekseySayarkin) [reply persistence/src/main/java/com/epam/esm/dao/GiftCertificateDAO.java:9]: Yes

File: persistence/src/main/java/com/epam/esm/dao/GiftCertificateDAO.java (hunk 3 above)
Comment (AlekseySayarkin) [reply persistence/src/main/java/com/epam/esm/dao/GiftCertificateDAO.java:12]: Done

File: persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java (hunk 10 above)
Comment (AlekseySayarkin) [reply persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:125]: If addGiftCertificateToDB fails it returns 0

File: persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java (hunk 11 above)
Comment (AlekseySayarkin) [reply persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:129]: Done

File: persistence/src/main/java/com/epam/esm/dao/extractor/GiftCertificateExtractor.java (hunk 6 above)
Comment (AlekseySayarkin) [reply persistence/src/main/java/com/epam/esm/dao/extractor/GiftCertificateExtractor.java:23]: Redesigned

File: persistence/src/main/java/com/epam/esm/dao/extractor/GiftCertificateExtractor.java (hunk 7 above)
Comment (AlekseySayarkin) [reply persistence/src/main/java/com/epam/esm/dao/extractor/GiftCertificateExtractor.java:30]: Changed them to final fields

File: persistence/src/main/java/com/epam/esm/dao/extractor/GiftCertificateExtractor.java (hunk 5 above)
Comment (AlekseySayarkin) [reply persistence/src/main/java/com/epam/esm/dao/extractor/GiftCertificateExtractor.java:42]: Renamed method name

File: persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java (hunk 8 above)
Comment (AlekseySayarkin) [reply persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:64]: removed create date from sql statement

File: persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java (hunk 12 above)
Comment (AlekseySayarkin) [reply persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:189]: Done

File: web/src/main/java/com.epam.esm/config/AppConfig.java (hunk 14 above)
Comment (AlekseySayarkin) [reply web/src/main/java/com.epam.esm/config/AppConfig.java:21]: Comments will be added in the last stage of development of Persistence module

File: persistence/src/main/java/com/epam/esm/dao/exception/PersistenceException.java [hunk 15]
```diff
@@ -0,0 +1,19 @@
+package com.epam.esm.dao.exception;
+
+public class PersistenceException extends Exception {
+
+    public PersistenceException() {
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/exception/PersistenceException.java:5]: Where is error code?

File: persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java [hunk 16]
```diff
@@ -0,0 +1,194 @@
...
+
+    private static final String SQL_GET_ALL_CERTIFICATES =
+            "select * from GiftCertificates " +
+            "left join CertificateDetails " +
+            "on GiftCertificates.ID = CertificateDetails.CertificateID " +
+            "left join Tags " +
+            "on Tags.ID = CertificateDetails.TagID ";
+
+    private static final String SQL_GET_CERTIFICATE_BY_TAG_NAME =
+            "select * from GiftCertificates " +
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:42]: Please not use exists

File: persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+    @Override
+    public List<GiftCertificate> getAllGiftCertificates() {
+        return jdbcTemplate.query(SQL_GET_ALL_CERTIFICATES, extractor);
+    }
+
+    @Override
+    public List<GiftCertificate> getAllGiftCertificates(String content) {
+        List<GiftCertificate> certificates = getAllGiftCertificates();
+
+        certificates.removeIf(giftCertificate -> !(giftCertificate.getName().contains(content) ||
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:104]: Please filter content in sql query

File: persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+            params[4] = giftCertificate.getDuration();
+        }
+
+        params[5] = giftCertificate.getId();
+
+        return params;
+    }
+
+    @Override
+    public boolean createJoin(int certificateId, int tagId) {
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:186]: Please rename method

File: persistence/src/main/java/com/epam/esm/dao/util/Utils.java [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+    private static final String CERTIFICATE_DESCRIPTION_COLUMN = "Description";
+    private static final String CERTIFICATE_PRICE_COLUMN = "Price";
+    private static final String CERTIFICATE_CREATE_DATE_COLUMN = "CreateDate";
+    private static final String CERTIFICATE_LAST_UPDATE_DATE_COLUMN = "LastUpdateDate";
+    private static final String CERTIFICATE_DURATION_COLUMN = "Duration";
+
+    private static final String TAG_ID_COLUMN = "TAGS.ID";
+    private static final String TAG_NAME_COLUMN = "TAGS.Name";
+
+    private Utils(){
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/util/Utils.java:29]: Please rename utils class. It`s not clear enough what is the purpose

File: persistence/src/test/java/com/epam/esm/dao/SQLGiftCertificateDaoImplTest.java [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+
+    @Test
+    public void whenAddGiftCertificate_thenCorrectlyReturnsIt() throws PersistenceException {
+        GiftCertificate added = initCertificate();
+        added.setId(giftCertificateDAO.addGiftCertificate(added));
+
+        giftCertificateDAO.createJoin(added.getId(), tagDao.getTagByName("tourism").getId());
+        giftCertificateDAO.createJoin(added.getId(), tagDao.getTagByName("relax").getId());
+
+        Assert.assertEquals(added, giftCertificateDAO.getGiftCertificate(added.getName()));
```
Comment (churaalena) [inline persistence/src/test/java/com/epam/esm/dao/SQLGiftCertificateDaoImplTest.java:75]: Dao call should not be called in the assertion

File: persistence/src/main/java/com/epam/esm/dao/util/Utils.java (hunk 19 above)
Comment (AlekseySayarkin) [reply persistence/src/main/java/com/epam/esm/dao/util/Utils.java:29]: done

File: persistence/src/test/java/com/epam/esm/dao/SQLGiftCertificateDaoImplTest.java (hunk 20 above)
Comment (AlekseySayarkin) [reply persistence/src/test/java/com/epam/esm/dao/SQLGiftCertificateDaoImplTest.java:75]: Fixed

File: persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java (hunk 18 above)
Comment (AlekseySayarkin) [reply persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:186]: Done

File: persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java (hunk 17 above)
Comment (AlekseySayarkin) [reply persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:104]: done

File: persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java (hunk 16 above)
Comment (AlekseySayarkin) [reply persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:51]: done

Comment (AlekseySayarkin) [reply persistence/src/main/java/com/epam/esm/dao/impl/SQLGiftCertificateDaoImpl.java:51]: Redesigned query to not use exist

File: persistence/src/main/java/com/epam/esm/dao/exception/PersistenceException.java (hunk 15 above)
Comment (AlekseySayarkin) [reply persistence/src/main/java/com/epam/esm/dao/exception/PersistenceException.java]: added error code

File: persistence/src/main/java/com/epam/esm/dao/extractor/GiftCertificateExtractor.java [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.List;
+
+public class GiftCertificateExtractor implements ResultSetExtractor<List<GiftCertificate>> {
+
+    @Override
+    public List<GiftCertificate> extractData(ResultSet rs) throws SQLException {
+        return com.epam.esm.dao.util.GiftCertificateExtractor.extractGiftCertificates(rs);
```
Comment (churaalena) [inline persistence/src/main/java/com/epam/esm/dao/extractor/GiftCertificateExtractor.java:14]: Please rename one of the util classes

File: persistence/src/test/java/com/epam/esm/dao/SQLTagDaoImpTest.java [hunk 22]
```diff
@@ -0,0 +1,83 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+public class SQLTagDaoImpTest {
+
+    private TagDao tagDao;
+
+    @Before
+    public void init() {
+        DataSource dataSource = new EmbeddedDatabaseBuilder()
```
Comment (churaalena) [inline persistence/src/test/java/com/epam/esm/dao/SQLTagDaoImpTest.java:24]: Test config can be extracted

**Review by churaalena — APPROVED** [review]

File: persistence/src/main/java/com/epam/esm/dao/extractor/GiftCertificateExtractor.java (hunk 21 above)
Comment (AlekseySayarkin) [reply persistence/src/main/java/com/epam/esm/dao/extractor/GiftCertificateExtractor.java:14]: Done

File: persistence/src/test/java/com/epam/esm/dao/SQLTagDaoImpTest.java (hunk 22 above)
Comment (AlekseySayarkin) [reply persistence/src/test/java/com/epam/esm/dao/SQLTagDaoImpTest.java:24]: Done

**Pull request merged**